In [1]:
import pandas as pd

import numpy as np
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier

In [2]:
#Read the file
df=pd.read_csv("Zoo.csv")

In [3]:
df

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0      aardvark     1         0     0     1         0        0         1   
1      antelope     1         0     0     1         0        0         0   
2          bass     0         0     1     0         0        1         1   
3          bear     1         0     0     1         0        0         1   
4          boar     1         0     0     1         0        0         1   
..          ...   ...       ...   ...   ...       ...      ...       ...   
96      wallaby     1         0     0     1         0        0         0   
97         wasp     1         0     1     0         1        0         0   
98         wolf     1         0     0     1         0        0         1   
99         worm     0         0     1     0         0        0         0   
100        wren     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

In [4]:
x=df.iloc[:,1:17]
x

hair  feathers  eggs  milk  airborne  aquatic  predator  toothed  \
0       1         0     0     1         0        0         1        1   
1       1         0     0     1         0        0         0        1   
2       0         0     1     0         0        1         1        1   
3       1         0     0     1         0        0         1        1   
4       1         0     0     1         0        0         1        1   
..    ...       ...   ...   ...       ...      ...       ...      ...   
96      1         0     0     1         0        0         0        1   
97      1         0     1     0         1        0         0        0   
98      1         0     0     1         0        0         1        1   
99      0         0     1     0         0        0         0        0   
100     0         1     1     0         1        0         0        0   

     backbone  breathes  venomous  fins  legs  tail  domestic  catsize  
0           1         1         0     0     4     0         0        1  
1           1         1         0     0     4     1         0        1  
2           1         0         0     1     0     1         0        0  
3           1         1         0     0     4     0         0        1  
4           1         1         0     0     4     1         0        1  
..        ...       ...       ...   ...   ...   ...       ...      ...  
96          1         1         0     0     2     1         0        1  
97          0         1         1     0     6     0         0        0  
98          1         1         0     0     4     1         0        1  
99          0         1         0     0     0     0         0        0  
100         1         1         0     0     2     1         0        0  

[101 rows x 16 columns]

In [5]:
y=df.iloc[:,17]
y

0      1
1      1
2      4
3      1
4      1
      ..
96     1
97     6
98     1
99     7
100    2
Name: type, Length: 101, dtype: int64

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
#splitting into testing and training dataset
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25)


In [8]:
import numpy
from pandas import read_csv
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# Using GridSearchCV method finding the best k value

In [9]:

n_neighbors = numpy.array(range(2,40))
param_grid = dict(n_neighbors=n_neighbors)

In [10]:
model = KNeighborsClassifier()
grid = GridSearchCV(estimator=model, param_grid=param_grid)
grid.fit(x, y)

C:\Users\91701\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:666: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=5.
  warnings.warn(("The least populated class in y has only %d"


GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39])})

In [11]:
print(grid.best_score_)
print(grid.best_params_) #Hence k=3 id best value

0.93
{'n_neighbors': 3}


In [12]:

kfold=KFold(n_splits=10)

In [13]:
model = KNeighborsClassifier(n_neighbors=3)
results = cross_val_score(model, x_train, y_train, cv=kfold)

In [14]:
print(results.mean()) 

0.9089285714285713


In [15]:
model = KNeighborsClassifier(n_neighbors=3)
results = cross_val_score(model, x_test, y_test, cv=kfold)

In [16]:
print(results.mean()) #Here there is an overftting problem.

0.75


# We can try to change the n_splits and get the best accuracy

In [17]:
kfold=KFold(n_splits=20)

In [18]:
model = KNeighborsClassifier(n_neighbors=3)
results = cross_val_score(model, x_train, y_train, cv=kfold)

In [19]:
print(results.mean())

0.9041666666666668


In [20]:
model = KNeighborsClassifier(n_neighbors=3)
results = cross_val_score(model, x_test, y_test, cv=kfold)

In [21]:
print(results.mean())  # here also there is an overfitting problem.

0.775


# Feature Extraction using Decision Tree(CART)

In [22]:
from sklearn.tree import  DecisionTreeClassifier
from sklearn import tree

In [23]:
from sklearn.tree import DecisionTreeClassifier
model_gini = DecisionTreeClassifier(criterion='gini', max_depth=3)

In [24]:
model_gini.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=3)

In [25]:
print(model_gini.feature_importances_)

[0.         0.26771399 0.         0.52671992 0.         0.
 0.         0.         0.         0.         0.         0.2055661
 0.         0.         0.         0.        ]


In [26]:
df.columns

Index(['animal name', 'hair', 'feathers', 'eggs', 'milk', 'airborne',
       'aquatic', 'predator', 'toothed', 'backbone', 'breathes', 'venomous',
       'fins', 'legs', 'tail', 'domestic', 'catsize', 'type'],
      dtype='object')

In [27]:
hair=pd.Series(df["hair"])
eggs=pd.Series(df["eggs"])
fins=pd.Series(df["fins"])
type=pd.Series(df["type"])

In [28]:
data=pd.DataFrame({'hair':hair,'eggs':eggs,'fins':fins,'type':type})

In [29]:
#Important features are extracted and formed DataFrame
data

hair  eggs  fins  type
0       1     0     0     1
1       1     0     0     1
2       0     1     1     4
3       1     0     0     1
4       1     0     0     1
..    ...   ...   ...   ...
96      1     0     0     1
97      1     1     0     6
98      1     0     0     1
99      0     1     0     7
100     0     1     0     2

[101 rows x 4 columns]

In [30]:
x=data.iloc[:,0:3]
x

hair  eggs  fins
0       1     0     0
1       1     0     0
2       0     1     1
3       1     0     0
4       1     0     0
..    ...   ...   ...
96      1     0     0
97      1     1     0
98      1     0     0
99      0     1     0
100     0     1     0

[101 rows x 3 columns]

In [31]:
y=data["type"]
y

0      1
1      1
2      4
3      1
4      1
      ..
96     1
97     6
98     1
99     7
100    2
Name: type, Length: 101, dtype: int64

In [32]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25)

In [33]:
kfold=KFold(n_splits=26)

In [34]:
model = KNeighborsClassifier(n_neighbors=2)
results = cross_val_score(model, x_train, y_train, cv=kfold)

In [35]:
print(results.mean())

0.5897435897435896


In [36]:
model = KNeighborsClassifier(n_neighbors=2)
results = cross_val_score(model, x_test, y_test, cv=kfold)

In [37]:
print(results.mean())

0.7692307692307693


# Hence we can conclude the this is our best model where accuracy for test=.589 and train data is .769 and there is no overfitting probelm also